### PR TITLE
used redis cache to get available currencies and when converting

### PR DESCRIPTION
**Pull request by Wamuyukariuki**

### CODE REVIEW

File: transactions/views.py [hunk 1]
```diff
@@ -26,6 +28,7 @@ def fetch_data_from_api(url):
         logger.error("Invalid JSON response from API")
         raise
 
+# Transaction creation with Redis cache for exchange rates
 class TransactionCreateView(generics.CreateAPIView):
     queryset = Transaction.objects.all()
     serializer_class = TransactionSerializer
```
Comment (abedkiloo) [inline transactions/views.py:43]: Use serializer validation of input don't do manual validation

File: transactions/views.py [hunk 2]
```diff
@@ -44,19 +47,26 @@ def create(self, request, *args, **kwargs):
...
-            if exchange_rate is None:
-                raise ValueError("Exchange rate missing in API response")
-        except Exception as e:
-            logger.error(f"Exchange rate API error: {str(e)}")
-            return Response({"error": "Failed to fetch exchange rate"}, status=status.HTTP_502_BAD_GATEWAY)
+        # Generate a unique cache key for the currency pair
+        cache_key = f"exchange_rate_{input_currency.upper()}_{output_currency.upper()}"
+        exchange_rate = cache.get(cache_key)
+
+        if exchange_rate is None:
```
Comment (abedkiloo) [inline transactions/views.py]: We should check exchange rate first before calling the API on line 48 you have already called the API right?

File: transactions/views.py (hunk 1 above)
Comment (Wamuyukariuki) [reply transactions/views.py:43]: sure I have updated

File: transactions/views.py (hunk 2 above)
Comment (Wamuyukariuki) [reply transactions/views.py]: right i have made the changes kindly review

**Review by abedkiloo — APPROVED** [review]

**Pull request merged**